File: easy/linked-list/python/234-palindrome-linked-list.py
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self) -> str:
        f = ""
        x = self
        while x is not None:
            f += f"{x.val}->"
            x = x.next
        return f


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        vals = []
        while head is not None:
            vals.append(head.val)
            head = head.next
        return vals == list(reversed(vals))


print(Solution().isPalindrome(ListNode(1, ListNode(2, ListNode(3)))))
print(Solution().isPalindrome(ListNode(1, ListNode(2, ListNode(1)))))
print(Solution().isPalindrome(ListNode(1)))
print(Solution().isPalindrome(None))

File: easy/linked-list/cpp/234-palindrome-linked-list.cpp

#include "../../../stdc++.h"
using std::vector;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        vector<int> vect;
        while (head != nullptr)
        {
            vect.push_back(head->val);
            head = head->next;
        }

        for (size_t i = 0; i < vect.size() / 2; i++)
        {
            int back = vect.size() - i - 1;
            if (vect[i] != vect[back])
            {
                return false;
            }
        }
        return true;
    }
};
